src/Servers/IIS/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp

```diff
@@ -319,9 +432,23 @@ FILE_WATCHER::StopMonitor()
     // we know that HandleChangeCompletion() call
     // can be ignored
     //
-    InterlockedExchange(&_lStopMonitorCalled, 1);
+    if (InterlockedExchange(&_lStopMonitorCalled, 1) == 1)
+    {
+        return;
+    }
+
+    LOG_INFO(L"Stopping file watching.");
+
     // signal the file watch thread to exit
     PostQueuedCompletionStatus(m_hCompletionPort, 0, FILE_WATCHER_SHUTDOWN_KEY, NULL);
+    WaitForMonitor(200);
+
+    if (m_fShadowCopyEnabled && m_pDoneCopyEvent != nullptr)
+    {
+        // If we are shadow copying, wait for the copying to finish.
+        WaitForSingleObject(m_pDoneCopyEvent, 30000);
```

Comment:
What is the behavior here if the 30 seconds completes and it's still copying?

Reply:
The app will finish shutting down. I could pass in the shutdown timeout here instead.

Reply:
I meant more of, will the app startup with a half copied directory?